### DISCUSSION TITLE
High memory consumption on Azure Function on Linux plan

### DISCUSSION THREAD

**Discussion by kmadof** [AzureCommunityPL/group#426] [Pomoc]
Niedawno przerzuciłem jedną funkcję z planu Windowsowego (p1v2) nad Linuksowy (p1v3). I dziwne rzeczy zaczęły się dziać z pamięcią.

![obraz](https://user-images.githubusercontent.com/1680172/165510193-f96171a8-13e0-477e-ae3a-cd0d57a17a97.png)

- kod jest ten sam
- kod jest budowany w ten sam sposób `dotnet publish -c Release --no-build --no-restore`
- funkcja jest w wersji ~3 i dotnet 3.1
- jest tam kilka funkcji s triggerem service busowym oraz timerem
- dedicated plan

Sam przebieg nasuwa na myśl wyciek pamięci, jednak ciężko mi to z czymkolwiek powiązać bo na windowsie takich akcji nie widać. 

Czy spotkaliście się z czymś takim? Co może być przyczyną tego zachowania?

**Comment by mmisztal1980** [comment]
W jakich trybach działa GC na:
- windows ?
- linux ?

Domyslam się, że w tym samym, ale chcę wykluczyć tą ścieżkę

**Reply by kmadof** [reply]
Nie było żadnych zmian w tym obszarze. W samym kodzie nie widzę ustawień dotyczących GC. Szukam sposobu, aby zweryfikować jaki tryb mam na planie windowsowym (na innej funkcji) oraz na linuksowym.

**Reply by kmadof** [reply]
@mmisztal1980 Miałeś nosa. Zmian nie było, ale:

- windows - workstation
- linux - server

Jutro sobie ustawię workstation na Linuxie i zobaczę jak to się zachowa.

**Reply by mmisztal1980** [reply]
Bhawo ja!

**Reply by mmisztal1980** [reply]
o ile pamiętam, server rezerwuje znacznie większe bloki pamięci niż workstation - stąd różnica

**Reply by mmisztal1980** [reply]
@kmadof https://docs.microsoft.com/en-us/dotnet/core/runtime-config/garbage-collector
tutaj masz całą dokumentację na temat ustawiania trybów GC. 
Osobiście polecam zmienne środowiskowe. ;)

**Reply by kaluzaaa** [reply]
@kmadof przypominam o tym guziku ;-)
![image](https://user-images.githubusercontent.com/4700951/165612255-d547fb77-6d43-4feb-8e95-c2aa7f2454a0.png)

**Reply by kmadof** [reply]
@kaluzaaa pamiętam o tym guziku, ale jeszcze nie udało mi się potwierdzić, że to ta kwestia. 

Póki co próbowałem przez zmienne środowiskowe. Apka w tym momencie siedzi na .NET 6 i tym samym ustawiałem `DOTNET_gcServer` na `0` jednak `System.Runtime.GCSettings.IsServerGC` dalej jest na `true`, a po samej pamięci nie widać różnic w zachowaniu.

[Tutaj znalazłem takie info](https://docs.microsoft.com/en-us/answers/questions/728091/azure-functions-garbage-collection-mode.html)
>1. Is there a way to do Server GC instead of workstation GC in Isolated Azure Functions?
No, the available configuration options for Azure Function is configured through host.json file and its reference is [available here.](https://docs.microsoft.com/en-us/azure/azure-functions/functions-host-json). It contains configuration that applies "after" the runtime has initialized. The GC mode settings must be configured "before" the dotnet runtime initializes. Furthermore, just changing the GC mode to server will not lower the size of memory. In-fact, in most scenarios the memory requirement increases because in server mode you have 1 dotnet heap per processor core. Setting the mode to server will benefit the applications which is intended to run as servers and need high throughput and scalability (not lower memory).
In the current scenario, I suspect that the web crawling is creating large objects (probably strings/XML docs) which are rooted in your application.

Tyle, że to się tyczy Isolated, a nie InProcess (w kontekście wybory GC mode).

**Comment by stic** [comment]
Nie w kodzie w konfigu
Cos a'la <ServerGarbageCollection>false</ServerGarbageCollection>
https://blog.markvincze.com/troubleshooting-high-memory-usage-with-asp-net-core-on-kubernetes/

Ale ostatnio mialem tez takiego kwiatka, wiec 'wszystko sie moze zdarzyc'..
https://github.com/dotnet/runtime/issues/13702

**Reply by kmadof** [reply]
@stic  I jak to się skończyło? Ustawiałeś `MALLOC_TRIM_THRESHOLD_ `?

**Comment by kamil-mrzyglod** [comment]
Tak w ciemno to nic nie znajdziesz. Powodów może być wiele. Spróbuj w pierwszej kolejności zrobić dumpa i go przeanalizować - masz taką możliwość w ramach **Diagnose and solve problems** na poziomu Azure Function:
![image](https://user-images.githubusercontent.com/11725731/165555094-739cfa5b-b37f-49ac-aecf-3e1f93ea2e3e.png)

**Reply by kamil-mrzyglod** [reply]
Plus dodatkowo pytanie czy jak postawisz na nowo na czystej Function App czy będziesz miał taki sam problem. Lokalnie próbowałeś uruchomić pod Linuxem? Nie ma tam takich kwiatków?

**Reply by kmadof** [reply]
To jest dobry pomysł, żeby lokalnie odpalić funkcję i sprawdzić/ustawi c tutaj GC Mode i porównać zachowanie.

**Comment by piotrgwiazda** [comment]
Zdaje się, że zmieniłeś plan z P1V2 na P1V3 a to tak z 2x więcej RAM (i
CPU). Możliwe, że GC będzie ogólnie mniej agresywny bo wykorzystujesz mniej
% pamięci dostępnej.

śr., 27 kwi 2022, 17:37 użytkownik Maciek Misztal ***@***.***>
napisał:

**Reply by kmadof** [reply]
Tak to prawda. Pamięci jest więcej. Chciałbym jednak przekonać się czy to jest ten przypadek (mniej agresywne podejście), zanim zapakuję cały plan funkcjami.
